crates/sparrow-runtime/src/prepare/preparer.rs

```diff
@@ -24,55 +34,148 @@ pub enum Error {
+const PREPARE_VERSION: i64 = 0;
+
 pub struct Preparer {
     prepared_schema: SchemaRef,
-    time_column_name: String,
-    subsort_column_name: Option<String>,
+    table_config: Arc<TableConfig>,
     next_subsort: AtomicU64,
-    key_column_name: String,
     time_multiplier: Option<i64>,
+    object_stores: Arc<ObjectStoreRegistry>,
 }
 
 impl Preparer {
     /// Create a new prepare produce data with the given schema.
     pub fn new(
-        time_column_name: String,
-        subsort_column_name: Option<String>,
-        key_column_name: String,
+        table_config: Arc<TableConfig>,
         prepared_schema: SchemaRef,
         prepare_hash: u64,
         time_unit: Option<&str>,
+        object_stores: Arc<ObjectStoreRegistry>,
     ) -> error_stack::Result<Self, Error> {
         let time_multiplier = time_multiplier(time_unit)?;
         Ok(Self {
             prepared_schema,
-            time_column_name,
-            subsort_column_name,
+            table_config,
             next_subsort: prepare_hash.into(),
-            key_column_name,
             time_multiplier,
+            object_stores,
         })
     }
 
     pub fn schema(&self) -> SchemaRef {
         self.prepared_schema.clone()
     }
 
+    /// Prepare a parquet file.
+    ///
+    /// - This computes and adds the key columns.
+    /// - This sorts the files by time, subsort and key hash.
+    /// - This adds or casts columns as needed.
+    /// - This produces multiple parts if the input file is large.
+    /// - This produces metadata files alongside data files.
+    pub async fn prepare_parquet(
+        &self,
+        path: &std::path::Path,
+    ) -> error_stack::Result<Vec<PreparedFile>, Error> {
+        // TODO: Support Slicing
+
+        // Prepared files are stored in the following format:
+        // file:///<cwd>/tables/<table_uuid>/prepared/<prepare_version>/<uuid>/part-<n>.parquet
+        let cur_dir = std::env::current_dir().expect("current dir");
+        let cur_dir = cur_dir.to_string_lossy();
+
+        let uuid = Uuid::new_v4();
+        let output_path_prefix = format!(
+            "file:///{}/tables/{}/prepare/{PREPARE_VERSION}/{uuid}/",
+            cur_dir, self.table_config.uuid
+        );
+        let output_file_prefix = "part";
+
+        let output_url = ObjectStoreUrl::from_str(&output_path_prefix)
+            .change_context_lazy(|| Error::InvalidUrl(path.to_string_lossy().to_string()))?;
+
+        let object_store = self
+            .object_stores
+            .object_store(&output_url)
+            .change_context(Error::Internal)?;
+
+        let source_data = SourceData {
+            source: Some(
+                SourceData::try_from_local(path)
+                    .into_report()
+                    .change_context(Error::Internal)?,
+            ),
+        };
+
+        // TODO: Slicing
+        let mut prepare_stream =
+            prepared_batches(&self.object_stores, &source_data, &self.table_config, &None)
```

Comment:
How does this size the file? We may want to do some batching or something, so that multiple prepared batches (or even all prepared batches) become part of the same prepared file.

Reply:
Yeah, it looks like prepare currently just makes a new file per batch. Probably agree we'd want to batch up batches to some desired size. 

But, seems outside of this PR, so will add a task to follow up with that.

Reply:
It uses.. this? 
```
fn get_num_files(file_size: usize) -> usize {
    // The number of files is the ceiling of the number of gigabytes in the file
    // e.g. 2.5 gb -> 3 files or 1 gb -> 1 file
    (file_size + GIGABYTE_IN_BYTES - 1) / GIGABYTE_IN_BYTES
}

fn get_batch_size(num_rows: usize, num_files: usize) -> usize {
    // To get the files ~1GB in size, we assume that the ceiling of the total number
    // of rows divided by the number of files.
    (num_rows + num_files - 1) / num_files
}
```

Reply:
Afaict, the batch size should be a gb, unless I misunderstand.